sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSourceReadSuite.scala

```diff
@@ -796,4 +973,228 @@ abstract class StateDataSourceReadSuite extends StateDataSourceTestBase with Ass
+    }
+  }
+
+  private def testSnapshotStateDfAgainstStateDf(resourceDir: File): Unit = {
+    val stateSnapshotDf = spark.read.format("statestore")
+      .option("snapshotPartitionId", 0)
+      .option("snapshotStartBatchId", 1)
+      .load(resourceDir.getAbsolutePath)
+
+    val stateDf = spark.read.format("statestore")
+      .load(resourceDir.getAbsolutePath)
+      .filter(col("partition_id") === 0)
+
+    checkAnswer(stateSnapshotDf, stateDf)
+  }
+
+  protected def testSnapshotOnLimitState(providerName: String): Unit = {
+    /** The golden files are generated by:
+    withSQLConf({
+      SQLConf.STREAMING_MAINTENANCE_INTERVAL.key -> "100"
+    }) {
+      val inputData = MemoryStream[(Int, Long)]
+      val query = inputData.toDF().limit(10)
+      testStream(query)(
+        StartStream(checkpointLocation = <...>),
+        AddData(inputData, (1, 1L), (2, 2L), (3, 3L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (4, 4L), (5, 5L), (6, 6L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (7, 7L), (8, 8L), (9, 9L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (10, 10L), (11, 11L), (12, 12L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) }
+      )
+    }
+     */
+    val resourceUri = this.getClass.getResource(
+      s"/structured-streaming/checkpoint-version-4.0.0/$providerName/limit/"
+    ).toURI
+
+    testSnapshotStateDfAgainstStateDf(new File(resourceUri))
+  }
+
+  protected def testSnapshotOnAggregateState(providerName: String): Unit = {
+    /** The golden files are generated by:
+    withSQLConf({
+      SQLConf.STREAMING_MAINTENANCE_INTERVAL.key -> "100"
+    }) {
+      val inputData = MemoryStream[(Int, Long)]
+      val query = inputData.toDF().groupBy("_1").count()
+      testStream(query, OutputMode.Update)(
+        StartStream(checkpointLocation = <...>),
+        AddData(inputData, (1, 1L), (2, 2L), (3, 3L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (2, 2L), (3, 3L), (4, 4L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (3, 3L), (4, 4L), (5, 5L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) },
+        AddData(inputData, (4, 4L), (5, 5L), (6, 6L)),
+        ProcessAllAvailable(),
+        Execute { _ => Thread.sleep(2000) }
+      )
+    }
+     */
+    val resourceUri = this.getClass.getResource(
+      s"/structured-streaming/checkpoint-version-4.0.0/$providerName/dedup/"
+    ).toURI
+
+    testSnapshotStateDfAgainstStateDf(new File(resourceUri))
+  }
+
+  protected def testSnapshotOnDeduplicateState(providerName: String): Unit = {
+    /** The golden files are generated by:
+    withSQLConf({
```

Comment:
nit: indent seems odd in these places, but maybe not a big deal for such comments

Reply:
Will move one tab right.